### PR TITLE
Wpf + save/load system

### PR DESCRIPTION
**Pull request by KarolDenst**
The main parts of the the system are:
Settings class that stores save locations, save name and seed.
Close event. On this event classes are supposed to save the data required to rebuild the scene.
Start event. On this event classes load saved data.

### CODE REVIEW

File: Hyper/Scene.cs [hunk 1]
```diff
@@ -141,7 +108,7 @@ private Camera GetCamera(float aspectRatio, float elevation)
         return camera;
     }
 
-    private PhysicalCharacter CreatePhysicalHumanoid(Vector3 initialPosition)
+    public PhysicalCharacter CreatePhysicalHumanoid(Vector3 initialPosition)
```
Comment (AleksyBalazinski) [inline Hyper/Scene.cs:111]: couldn't is stay `private`?

File: Hyper/Window.cs [hunk 2]
```diff
@@ -1,92 +1,35 @@
...
-    private readonly Context _context = Common.UserInput.Context.Instance;
+    private static readonly Logger Logger = LogManager.GetCurrentClassLogger();
 
     public Window(GameWindowSettings gameWindowSettings, NativeWindowSettings nativeWindowSettings)
         : base(gameWindowSettings, nativeWindowSettings)
     {
-        StartDebugThreadAsync();
-
-        RegisterCallbacks();
+        _game = new Game(nativeWindowSettings.Size.X, nativeWindowSettings.Size.Y, new WindowHelper(this), Guid.NewGuid().ToString());
```
Comment (AleksyBalazinski) [inline Hyper/Window.cs:18]: Maybe use current time instead of GUID? It'd be super nice if the user could specify that as well

File: Common/Settings.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        Seed = seed;
+        SaveName = saveName;
+        AspectRatio = aspectRatio;
+
+        if (!Directory.Exists(AppDataLocation)) Directory.CreateDirectory(AppDataLocation);
+        if (!Directory.Exists(SavesLocation)) Directory.CreateDirectory(SavesLocation);
+        if (!Directory.Exists(CurrentSaveLocation)) Directory.CreateDirectory(CurrentSaveLocation);
+    }
+
+    public static Settings Load(string saveName, float aspectRatio)
```
Comment (AleksyBalazinski) [inline Common/Settings.cs:33]: consider changing the signature so that we don't have to pass aspectRatio: it makes the impression like this is needed to load the settings info. Maybe something like `bool TryLoad(string saveName, out Settings)` ?

File: Hyper/Game.cs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+using OpenTK.Graphics.OpenGL4;
+using OpenTK.Mathematics;
+using OpenTK.Windowing.Common;
+using OpenTK.Windowing.GraphicsLibraryFramework;
+
+namespace Hyper;
+
+public class Game : IInputSubscriber
+{
+    public bool IsRunning = true;
```
Comment (AleksyBalazinski) [inline Hyper/Game.cs:17]: now when the window loses focus, the animations keep playing. Consider setting this to false when focus is lost

File: Hyper/Game.cs [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        GL.Viewport(0, 0, _size.X, _size.Y);
+        _scene.Camera.AspectRatio = _size.X / (float)_size.Y;
+        _size = e.Size;
+    }
+
+    public void RegisterCallbacks()
+    {
+        _context.RegisterKeys(new List<Keys> { Keys.Escape });
+
+        _context.RegisterKeyDownCallback(Keys.Escape, Close);
```
Comment (AleksyBalazinski) [inline Hyper/Game.cs:166]: hitting esc causes an exception to be thrown from ChunkWorker.RunJobs : `System.OperationCanceledException: 'The operation was canceled.'`

File: Chunks/ChunkManagement/ChunkWorker.cs [hunk 6]
```diff
@@ -41,30 +43,53 @@ private enum JobType
...
         for (int i = 0; i < NumberOfThreads; i++)
         {
             Task.Run(RunJobs);
         }
+
+        GetSavedChunks();
+        EnqueueLoadingChunks(Vector3i.Zero);
+        int totalChunks = (2 * RenderDistance + 1) * (2 * RenderDistance + 1) * (2 * RenderDistance + 1);
+        int prevNumber = 0;
+        while (_chunks.Count < totalChunks)
```
Comment (AleksyBalazinski) [inline Chunks/ChunkManagement/ChunkWorker.cs:84]: I'm not sure I get it. Could we discuss it later?

**Review by AleksyBalazinski — CHANGES_REQUESTED** [review]
This PR is rather huge & I didn't properly test everything, I'll spend more time on this & write more comments if I'll have more feedback. Also note I didn't review the wpf part.

File: Common/UserInput/Context.cs [hunk 7]
```diff
@@ -156,4 +162,28 @@ public void ExecuteAllHeldCallbacks(InputType inputType, FrameEventArgs e)
                 break;
         }
     }
+
+    public void Clear()
```
[the remark this answers is not in the document]
Comment (KarolDenst) [reply Common/UserInput/Context.cs:166]: we cannot. This is a singleton so we need to remove them when we save a game and then load it

[the remark this answers is not in the document]
Comment (KarolDenst) [reply Common/UserInput/Context.cs:166]: I'll look into other places with Clear()

File: Common/IWindowHelper.cs
[the remark this answers is not in the document]
Comment (KarolDenst) [reply Common/IWindowHelper.cs:1]: Because we need 2 WindowHelpers. one for wpf and one for openTK.Window

File: Hyper/Game.cs [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    {
+        _size = new Vector2i(width, height);
+
+        GL.ClearColor(0f, 0f, 0f, 1.0f);
+        GL.Enable(EnableCap.DepthTest);
+        GL.Enable(EnableCap.Blend);
+        GL.BlendFunc(BlendingFactor.SrcAlpha, BlendingFactor.OneMinusSrcAlpha);
+
+        _settings = Settings.Load(saveName, height / (float)width);
+        _scene = new Scene(_size.X / (float)_size.Y, 31);
```
[the remark this answers is not in the document]
Comment (KarolDenst) [reply Hyper/Game.cs:39]: RIght now the player position is initialized before the chunks are created so it's not possible. We will have to add a way to spawn bots anyways later on so i thought we would use that to get the player position as well when we develop it

File: Chunks/ChunkManagement/ChunkWorker.cs [hunk 9]
```diff
@@ -41,30 +43,53 @@ private enum JobType
...
-    private readonly CancellationTokenSource _cancellationTokenSource = new();
+    private CancellationTokenSource _cancellationTokenSource = new();
 
-    public ChunkWorker(List<Chunk> chunks, SimulationManager<PoseIntegratorCallbacks> simulationManager, ChunkFactory chunkFactory)
+    public ChunkWorker(List<Chunk> chunks, SimulationManager<PoseIntegratorCallbacks> simulationManager, Settings settings)
     {
         _chunks = chunks;
         _simulationManager = simulationManager;
-        _chunkFactory = chunkFactory;
+        _chunkFactory = new ChunkFactory(new ScalarFieldGenerator(settings.Seed));
```
[the remark this answers is not in the document]
Comment (KarolDenst) [reply Chunks/ChunkManagement/ChunkWorker.cs:62]: I don't know if we need to decuple these since I think that ChunkFactory shoould only be used by chunk worker and no other class

File: Chunks/ChunkManagement/ChunkWorker.cs [hunk 10]
```diff
@@ -41,30 +43,53 @@ private enum JobType
...
+    private void Start()
+    {
+        _cancellationTokenSource = new CancellationTokenSource();
         for (int i = 0; i < NumberOfThreads; i++)
         {
             Task.Run(RunJobs);
         }
+
+        GetSavedChunks();
+        EnqueueLoadingChunks(Vector3i.Zero);
```
[the remark this answers is not in the document]
Comment (KarolDenst) [reply Chunks/ChunkManagement/ChunkWorker.cs:81]: I wanted to move loading initial chunks from scene to chunkWoerker

File: Chunks/ChunkManagement/ChunkWorker.cs [hunk 11]
```diff
@@ -215,4 +240,52 @@ private static float GetDistance(Vector3i a, Vector3i b)
...
+        _cancellationTokenSource.Cancel();
+        _cancellationTokenSource.Dispose();
+
+        _existingChunks.Clear();
+        _chunksToLoad.Clear();
+        _loadedChunks.Clear();
+        _savedChunks.Clear();
+        _chunksToUpdateQueue.Clear();
+        _chunksToUpdateHashSet.Clear();
+        _updatedChunks.Clear();
```
[the remark this answers is not in the document]
Comment (KarolDenst) [reply Chunks/ChunkManagement/ChunkWorker.cs:280]: If i don't i have memory leaks. Will look into it

File: Chunks/ChunkManagement/ChunkWorker.cs (hunk 6 above)
Comment (KarolDenst) [reply Chunks/ChunkManagement/ChunkWorker.cs:84]: sure

File: Hyper/Game.cs (hunk 4 above)
Comment (KarolDenst) [reply Hyper/Game.cs:17]: will fix that. but also just for the future we have to stop using Stopwatch and just use UpdateFrame event times since we don't want the game state to change when it's paused/out of focus

File: Hyper/Game.cs (hunk 5 above)
Comment (KarolDenst) [reply Hyper/Game.cs:166]: I have not been able to reproduce this. Can you give some more details?

File: Common/WindowHelper.cs
[the remark this answers is not in the document]
Comment (KarolDenst) [reply Common/WindowHelper.cs:1]: because we need 2 windowHelpers. one for wpf one for opentk window

File: Chunks/ChunkManagement/ChunkWorker.cs [hunk 12]
```diff
@@ -215,4 +240,52 @@ private static float GetDistance(Vector3i a, Vector3i b)
...
+
+        SaveAllChunks();
```
[the remark this answers is not in the document]
Comment (KarolDenst) [reply Chunks/ChunkManagement/ChunkWorker.cs:286]: I don't really see a clean solution. How about changing Dispose() to Close() and just keeping the same logic

File: Hyper/Game.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (AleksyBalazinski) [reply Hyper/Game.cs:40]: I was thinking about using dependency injection here to make avgElevation available at scene creation, here's what I had in mind: https://github.com/Non-Euclidean-World/Hyper/commit/6a0534569c429a8adca9d55d9109f7f37f151582 (it was easier to code it than to describe here in plain English)

File: Chunks/ChunkManagement/ChunkWorker.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (AleksyBalazinski) [reply Chunks/ChunkManagement/ChunkWorker.cs:81]: Any updates on this one?

File: Chunks/ChunkManagement/ChunkWorker.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (AleksyBalazinski) [reply Chunks/ChunkManagement/ChunkWorker.cs:284]: Any updates on this one?

File: Chunks/ChunkManagement/ChunkWorker.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (AleksyBalazinski) [reply Chunks/ChunkManagement/ChunkWorker.cs:286]: We could make `SaveAllChunks` public and call it in `Game` trough `ChunkController.SaveAllChunks` which wraps that method.
But that's not really important

File: Common/UserInput/Context.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (AleksyBalazinski) [reply Common/UserInput/Context.cs:166]: > we cannot. This is a singleton so we need to remove them when we save a game and then load it

Right, so maybe this indicates that it shouldn't be a singleton in the first place? Can we consider making it a field in `Game` class and passing down to other classes as necessary?

[the remark this answers is not in the document]
Comment (KarolDenst) [reply Common/UserInput/Context.cs:166]: I like this idea. will implement it

File: Hyper/Game.cs (hunk 5 above)
Comment (KarolDenst) [reply Hyper/Game.cs:175]: I'm not getting this. are you using the latest version of this branch?

Comment (AleksyBalazinski) [reply Hyper/Game.cs:175]: To the best of my knowledge I am, steps to reproduce are literally the following:
1. Run Hyper project
2. Hit <kbd>Esc</kbd>
3.  `System.OperationCanceledException: 'The operation was canceled.'` (in ChunkWorker)

File: Chunks/ChunkManagement/ChunkWorker.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (KarolDenst) [reply Chunks/ChunkManagement/ChunkWorker.cs]: removed the clears. The momory leak seems to stop just like in chunkWorker

File: Hyper/Game.cs (hunk 5 above)
Comment (KarolDenst) [reply Hyper/Game.cs]: should be fixed now

Comment (AleksyBalazinski) [reply Hyper/Game.cs]: confirmed

File: WpfMenu/ExampleScene.cs
Comment (AleksyBalazinski) [inline WpfMenu/ExampleScene.cs:1]: How is this used?

**Review by AleksyBalazinski — CHANGES_REQUESTED** [review]
I left some comments. I can't vouch for the WPF part, but I tested it & it seems to work.

Also, a little afterthought about formatting: we're sometimes using the first style below and sometimes the second. It's quite hard to put breakpoints on the `if`'s body when the code is formatted like that
```
if (condition) something;
else something_else;
```
could we set a new rule so that it enforces
```
if (condition)
    something;
else
    something_else;
```
?

File: Hyper/Game.cs [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+        }
+        _scene.Dispose(); // Scene dispose needs to be after controller dispose.
+        _settings.Save();
+        LogManager.Flush();
+        IsRunning = false;
+    }
+
+    public void RenderFrame(FrameEventArgs e)
+    {
+        if (e.Time == 0) return;
```
[the remark this answers is not in the document]
Comment (KarolDenst) [reply Hyper/Game.cs:84]: in wpf after game is paused and than resumed

[the remark this answers is not in the document]
Comment (KarolDenst) [reply Hyper/Game.cs:84]: Since it only occurs in wpf i iwll move this check there

**Review by AleksyBalazinski — APPROVED** [review]
